daemon/src/rollover_taker.rs

```diff
@@ -0,0 +1,318 @@
+use crate::oracle::{self, GetAnnouncement};
+use crate::projection::{self, UpdateRollOverProposal};
+use crate::setup_contract::{self, RolloverParams};
+use crate::tokio_ext::spawn_fallible;
+use crate::wire::{self, RollOverMsg};
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use futures::channel::mpsc::{self, UnboundedSender};
+use futures::{future, SinkExt};
+use maia::secp256k1_zkp::schnorrsig;
+use xtra::prelude::MessageChannel;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    cfd: Cfd,
+    n_payouts: usize,
+    oracle_pk: schnorrsig::PublicKey,
+    timestamp: Timestamp,
+    maker: xtra::Address<connection::Actor>,
+    get_announcement: Box<dyn MessageChannel<GetAnnouncement>>,
+    projection: xtra::Address<projection::Actor>,
+    on_completed: Box<dyn MessageChannel<Completed>>,
+    on_stopping: Vec<Box<dyn MessageChannel<Stopping<Self>>>>,
+    rollover_msg_sender: Option<UnboundedSender<RollOverMsg>>,
+}
+
+impl Actor {
+    pub fn new(
+        (cfd, n_payouts): (Cfd, usize),
+        oracle_pk: schnorrsig::PublicKey,
+        maker: xtra::Address<connection::Actor>,
+        get_announcement: &(impl MessageChannel<GetAnnouncement> + 'static),
+        projection: xtra::Address<projection::Actor>,
+        on_completed: &(impl MessageChannel<Completed> + 'static),
+        (on_stopping0, on_stopping1): (
+            &(impl MessageChannel<Stopping<Self>> + 'static),
+            &(impl MessageChannel<Stopping<Self>> + 'static),
+        ),
+    ) -> Self {
+        Self {
+            cfd,
+            n_payouts,
+            oracle_pk,
+            timestamp: Timestamp::now(),
+            maker,
+            get_announcement: get_announcement.clone_channel(),
+            projection,
+            on_completed: on_completed.clone_channel(),
+            on_stopping: vec![on_stopping0.clone_channel(), on_stopping1.clone_channel()],
+            rollover_msg_sender: None,
+        }
+    }
+
+    async fn propose(&self, this: xtra::Address<Self>) -> Result<()> {
+        self.maker
+            .send(connection::ProposeRollOver {
+                order_id: self.cfd.order.id,
+                timestamp: self.timestamp,
+                address: this,
+            })
+            .await??;
+
+        self.update_proposal(Some((
+            RollOverProposal {
+                order_id: self.cfd.order.id,
+                timestamp: self.timestamp,
+            },
+            SettlementKind::Outgoing,
+        )))
+        .await?;
+
+        Ok(())
+    }
+
+    async fn handle_confirmed(
+        &mut self,
+        msg: RollOverAccepted,
+        ctx: &mut xtra::Context<Self>,
+    ) -> Result<()> {
+        let RollOverAccepted { oracle_event_id } = msg;
+        let announcement = self
+            .get_announcement
+            .send(oracle::GetAnnouncement(oracle_event_id))
+            .await?
+            .with_context(|| format!("Announcement {} not found", oracle_event_id))?;
```

Comment:
I guess here we can't avoid it so could potentially include this in the other protocols as well? But then the question is, where do you draw the line? Should each of these actors perhaps also load the CFD from the DB themselves? :thinking:

Reply:
Perhaps not such a bad idea actually because then even more error cases are covered ...

Reply:
I was leaning towards even loading from the `db` originally. I'm really not sure what the CFD actors should be doing other than handling user actions. Even handling completion of these ephemeral protocols may be unnecessary.

Reply:
> I'm really not sure what the CFD actors should be doing other than handling user actions.

Someone needs to track the currently on-going protocols to make sure we are not executing them in parallel for the same CFD. In theory, we could overcome this (once we have event-sourcing) by recording the _expected_ version of the CFD in the protocol and if you try to apply the result to a different one, it would fail (and it means you'd have to retry). But if you do that, how would you communicate this back to the user / the original interaction? Thinking about it more though, it is really almost impossible to trigger multiple of these through the UI.

We could take the approach of not guarding against this in the API and let it only fail eventually with a log message in case someone manages to somehow trigger this?

> Even handling completion of these ephemeral protocols may be unnecessary.

Once we do event-sourcing, I imagine these to go away yes because we will simply broadcast the resulting event of applying the completed state to the CFD.

Reply:
To proceed, shall we just leave it as it is and iterate on all this as we go?

Reply:
Yes, leave as is for now. Mostly something to mull over :)